feat: add burst (super) reactions

spifory opened the pull request:
## Summary

- Implements discord/discord-api-docs#6056 & discord/discord-api-docs#6375

Tested with: https://paste.nextcord.dev/?id=1725902791782971

![image](https://github.com/user-attachments/assets/0105c79f-a27e-4973-a1fb-d51710c49a16)
![image](https://github.com/user-attachments/assets/bbe1fdd3-ef8a-4ea5-9878-828a9292f1bb)
![image](https://github.com/user-attachments/assets/e5ddf220-fd02-4109-aaca-ca6f1e8096c7)

I'm still uncertain whether or not bots *can* or cannot send burst reactions, however the docs don't document anything about it

<!-- Link any issues if applicable.
Use keywords from https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests
-->

<!-- Uncomment based on the type of your changes below -->


## This is a **Code Change**

- [x] I have tested my changes.
- [x] I have updated the documentation to reflect the changes.
- [x] I have run `task pyright` and fixed the relevant issues.

spifory committed:
feat: add burst (super) reactions

ooliver1 commented on nextcord/reaction.py:
```diff
@@ -2,18 +2,42 @@
 
 if TYPE_CHECKING:
     from .abc import Snowflake
+    from .colour import Color
```
From a glance, looks like this is in type checking but used below for `burst_colors` (don't we use british english). For that anyway, it could probably be a (cached) property to save memory for the majority of people that do not want the burst reaction colours.

spifory replied:
Oh okay thanks, will do that
Should I replace `burst_colors` with colours to make an alias?

spifory committed:
feat: handle burst reactions through events

spifory committed:
typing(reaction)!: use Colour

spifory committed:
fix(reaction): remove double slot

spifory committed:
refactor(reaction): add burst_colours and burst_colors aliases

spifory committed:
refactor(reaction): make burst_colours a cached_property

teaishealthy commented on nextcord/reaction.py:
```diff
@@ -2,18 +2,43 @@
+    __slots__ = ("burst", "normal")
+
+    def __init__(self, data: ReactionCountDetailsPayload) -> None:
+        self.burst = data.get("burst")
+        self.normal = data.get("normal")
```
types?

teaishealthy replied:
```suggestion
        self.burst: int = data.get("burst")
        self.normal: int = data.get("normal")
```

EmmmaTech commented on nextcord/raw_models.py:
```diff
@@ -168,7 +169,17 @@ class RawReactionActionEvent(_RawReprMixin):
+        "event_type",
+        "member",
+        "burst",
+        "burst_colors",
```
There seems to be a missing [type attribute](https://discord.com/developers/docs/resources/message#get-reactions-reaction-types) here, although I'm not sure why it exists since the `burst` boolean is also there...

EmmmaTech commented on nextcord/reaction.py:
```diff
@@ -92,6 +139,22 @@ def __str__(self) -> str:
 
+    @cached_property
+    def burst_colours(self) -> List[Colour]:
+        """List[:class:`Colour`]: The HEX colors used for a burst reaction.
```
I'm confused on whether we're sticking with American or British English here. Also why is hex in uppercase?

alentoghostflame replied:
Traditionally, there is a function with the British English name that holds the functionality and an American alias for it. An example would be [User.colour and User.color](https://github.com/nextcord/nextcord/blob/master/nextcord/user.py#L215-L231).

EmmmaTech replied:
I'm aware of this, but the problem is American English spelling is used in this docstring. In the example you brought up, the proper spellings for the dialects are used.

spifory replied:
My mistake British English it is. Also "HEX" is capitalised for no reason in particular really, AFAIK you can write it capitalised or uncapitalised

spifory committed:
Merge branch 'master' into feat/add-super-reactions

spifory committed:
fix(reaction): use cached_slot_property

spifory committed:
refactor(reaction): explicitly type `burst` and `normal`

spifory committed:
docs(reaction): british "colors"

spifory committed:
feat: add `Reaction.type`

spifory committed:
fix(reaction): add _cs_burst_colours slot

spifory committed:
fix(reaction): remove double burst_colours slot

spifory committed:
docs(reaction): remove trailing backtick

spifory committed:
fix(reaction)!: add _cs_burst_colours to slots

spifory committed:
fix: add back _burst_colours

spifory committed:
Merge branch 'master' into feat/add-super-reactions

spifory committed:
fix(reaction): move Colour out of TYPE_CHECKING block

spifory committed:
fix(reaction): burst_colours is not an integer

Discord can't be consistent for shit apparently. (assumed it functioned like role colours.)

spifory committed:
fix(http): use reaction type value

spifory committed:
feat: add `type` to Reaction.users function

Had no idea this function existed until now

ooliver1 commented on nextcord/reaction.py:
```diff
@@ -13,7 +16,30 @@
+    normal: :class:`int`
+        The amount of normal reactions.
+
+    .. versionadded:: 3.0
```
This goes before attributes, after any main description.

spifory committed:
Merge branch 'master' into feat/add-super-reactions

spifory committed:
docs(enums): learn how to spell

spifory committed:
docs(reaction): move versionadded position

spifory committed:
docs(raw_models): document `burst`

spifory committed:
fix(types): remove `burst` from ReactioClearEmoji

EmmmaTech reviewed (changes requested):
tested, and for newly created messages with a new reaction, `count_details` seems to be missing when provided from the api. I was able to get my bot to crash a few times due to this.
here's my test code if necessary: https://paste.nextcord.dev/?id=1732326138949174
exception traceback if necessary: https://paste.nextcord.dev/?id=1732326348787843

teaishealthy commented on nextcord/reaction.py:
```diff
@@ -13,7 +16,30 @@
+    __slots__ = ("burst", "normal")
+
+    def __init__(self, data: ReactionCountDetailsPayload) -> None:
+        self.burst: int = data.get("burst")
+        self.normal: int = data.get("normal")
```
doesn't have to be .get()

teaishealthy commented on nextcord/reaction.py:
```diff
@@ -69,6 +112,12 @@ def __init__(
         self.me: bool = data.get("me")
+        self.me_burst: bool = data.get("me_burst")
+        self.count_details: ReactionCountDetails = ReactionCountDetails(
+            data=data.get("count_details")
```
same here

teaishealthy reviewed (commented):
okay™

spifory committed:
Merge branch 'master' into feat/add-super-reactions

spifory committed:
refactor(reaction): use direct key access instead of .get

spifory committed:
docs: change versionadded to 3.2

spifory committed:
Merge branch 'master' into feat/add-super-reactions

ooliver1 committed:
Merge branch 'master' into feat/add-super-reactions

teaishealthy reviewed (approved):
untested

ooliver1 committed:
Merge branch 'master' into feat/add-super-reactions

ooliver1 commented on nextcord/types/raw_models.py:
```diff
@@ -35,6 +36,9 @@ class ReactionActionEvent(TypedDict):
     member: NotRequired[Member]
+    burst: bool
+    burst_colors: NotRequired[List[str]]
+    type: ReactionType
```
The type here isn't an enum, its a `Literal[0, 1]`, so needs `ReactionType = Literal[0, 1]` near the top of this file instead of an enums import

ooliver1 commented on nextcord/iterators.py:
```diff
@@ -9,17 +9,18 @@
-    "archived_thread_iterator",
+    "reaction_iterator",
     "scheduled_event_iterator",
     "scheduled_event_user_iterator",
```
I'd leave these alone and leave sorting to a future PR (enabling the ruff rule for it ofc)

spifory replied:
Oh my fault     I likely hit save on the file back then and it formatted that too

ooliver1 commented on nextcord/raw_models.py:
```diff
@@ -148,6 +150,10 @@ class RawReactionActionEvent(_RawReprMixin):
+        If the reaction is a super/burst reaction.
+
+        .. versionadded:: 3.2
 
```
+ document the `type` attribute

ooliver1 commented on nextcord/http.py:
```diff
@@ -1612,6 +1612,7 @@ def get_reaction_users(
         emoji: str,
         limit: int,
         after: Optional[Snowflake] = None,
+        type: Optional[ReactionType] = None,
```
It seems like you can also pass `type` as a query param to `add_reaction`
![image.png](https://github.com/user-attachments/assets/f8a83e41-e7e7-4919-89ba-ab8f14dda26a)
